### PR TITLE
Migrate to ginkgo v2 

### LINKED ISSUE #5791: Upgrade ginkgo to version 2.0
## Which functionality do you think we should add?
Ginkgo 2.0 is now available and offers additional reporting functionality

## Why is this needed?
Upgrading to Ginkgo v2.0 offers: 
- new GinkgoWriter methods to write output that is only made visible if a test fails 
- improved reporting (machine-readable reports, generation of custom reports) ability to control output

An [upgrade guide and new feature list ](https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md) is available.

### PR DESCRIPTION
**Pull request by anandrkskd**
**What type of PR is this:**

/kind tests

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**
Fixes #5791

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: tests/integration/cmd_namespace_test.go
Comment (feloy) [inline tests/integration/cmd_namespace_test.go:181]: We should avoid using this, as it will result in more flaky tests, and so much longer execution times
I proposed another solution in https://github.com/redhat-developer/odo/pull/5837

File: tests/e2escenarios/e2e_devfile_test.go
Comment (feloy) [inline tests/e2escenarios/e2e_devfile_test.go:67]: We should avoid using this IMO

**Review by feloy — CHANGES_REQUESTED** [review]

File: tests/integration/cmd_namespace_test.go
Comment (anandrkskd) [reply tests/integration/cmd_namespace_test.go]: Understood! Will revert these changes.

File: Makefile
Comment (rm3l) [inline Makefile:119]: Just curious, using the vendored version no longer works with this upgrade?
While I understand the previous behavior of running Ginkgo from the vendor folder (making sure the right version of Ginkgo is running), I think installing the `latest` version might be dangerous.

Comment (anandrkskd) [reply Makefile:119]: When we are trying to use it the old way using vendor, it was not recognizing the new flags, and was not honoring the old flags.
I think we can use vendor after we merge [PR](https://github.com/redhat-developer/odo/pull/5833) in.

Comment (anandrkskd) [reply Makefile:119]: I will update that with a specific version, thanks.

Comment (anandrkskd) [reply Makefile]: I think it will make this PR huge. I can create a new PR replacing your #5833. We can merge this PR after the cleanup PR if merged?
WDYT @feloy

**Comment by anandrkskd** [comment]
@feloy the validate test is failing with 
```go run -mod=vendor github.com/frapposelli/wwhrd check -q
ERRO[0001] Found Non-Approved license      license=UNKNOWN package=github.com/segmentio/backo-go
FATA[0001] Exiting: Non-Approved license found          
exit status 1
```
And this used inside segmentio package, Not sure what should be my next step.

**Comment by feloy** [comment]
> @feloy the validate test is failing with
> 
> ```go
> ERRO[0001] Found Non-Approved license      license=UNKNOWN package=github.com/segmentio/backo-go
> FATA[0001] Exiting: Non-Approved license found          
> exit status 1
> ```
> 
> And this used inside segmentio package, Not sure what should be my next step.

This is because we are now including the v1.0.0 instead of at a specific commit:

```

File: scripts/generate-mod.sh
Comment (anandrkskd) [reply scripts/generate-mod.sh]: added

File: .ibm/pipelines/kubernetes-tests.sh
Comment (rm3l) [inline .ibm/pipelines/kubernetes-tests.sh:15]: Since we are still running Ginkgo from the `vendor` folder, maybe this is no longer necessary.

File: .ibm/pipelines/openshift-tests.sh
Comment (rm3l) [inline .ibm/pipelines/openshift-tests.sh:16]: Same comment: this is no longer necessary.

File: scripts/generate-mod.sh
Comment (rm3l) [inline scripts/generate-mod.sh:28]: Trying to understand the `why?`.. @feloy Can we elaborate a little bit on this? A comment in this file would be helpful for all of us..

File: scripts/openshiftci-presubmit-all-tests.sh
Comment (rm3l) [inline scripts/openshiftci-presubmit-all-tests.sh:19]: Same comment: Running `make goget-ginkgo` is no longer necessary, I think.

File: scripts/generate-mod.sh
Comment (rm3l) [inline scripts/generate-mod.sh:1]: @feloy Unless I've missed something about the usage of this script, I'm wondering if it wouldn't make sense to run this as part of CI, to detect cases where the `go.mod` file would be edited directly without updating this script (e.g, Dependabot PRs or someone updating/adding a new dependency). This way, we would be aware of editing this file if there is a diff.
But I guess it might not be straightforward especially where this script might potentially update dependencies with `go get -u`..

File: .ibm/pipelines/kubernetes-tests.sh
Comment (anandrkskd) [reply .ibm/pipelines/kubernetes-tests.sh:15]: removed

File: scripts/generate-mod.sh
Comment (feloy) [reply scripts/generate-mod.sh:28]: @dharmit I really don't know, that's why I added a why? here meaning "TODO: understand why `go mod tidy` fails and the option `-compat=1.17` is necessary".

File: scripts/generate-mod.sh
Comment (feloy) [reply scripts/generate-mod.sh:22]: IMO we cannot, because this one is not used directly in the odo sources, but in one of the dependencies

File: scripts/generate-mod.sh
Comment (feloy) [reply scripts/generate-mod.sh:1]: I agree that I would make sense. I remember we discussd to put in place a script to check these kind of generated code (mocks at least); that would be part of this work

File: scripts/generate-mod.sh
Comment (dharmit) [reply scripts/generate-mod.sh:28]: Ah, ok. I thought it was Anand who added the comment. My bad.

I was doing some reading around this after noticing the code comment and my response to Armel's comment was based on what I understood. I mainly referred to this [stackoverflow answer](https://stackoverflow.com/a/71974997/395670) and below excerpt from [Go 1.17 release notes](https://go.dev/doc/go1.17#go-command):
> The -compat flag allows that version to be overridden to support older (or only newer) versions, up to the version specified by the go directive in the go.mod file. To tidy a go 1.17 module for Go 1.17 only, without saving checksums for (or checking for consistency with) Go 1.16:

File: scripts/generate-mod.sh
Comment (rm3l) [reply scripts/generate-mod.sh:1]: Indeed, I also remember a discussion around such checks. I'll try to find a related issue (or create a new one) to track this.

**Review by rm3l — APPROVED** [review]

**Pull request closed without merging**
